### DISCUSSION TITLE
Building a multilingual site using wheels

### DISCUSSION THREAD

**Discussion by edthetechie** [wheels-dev/wheels#1252] [Q&A] [answered]
Hey guys,

We're looking to bid for some work that will involve building a site to support multiple languages, I can't find any wheels plugins or anything in the guide about this so I thought I'd ask how you guys are doing it? Have any of you got a multilingual site running and if so how do you handle the language files and things like that?

Any help much appreciated as always :-)

Thanks!

**Comment by dhgassoc** [comment]
I am not a wheels expert, but have made a few apps with cfwheels.  I have never done multiple languages.  A few questions;
a) is the data in the database in a single languages (ie. numbers or statistics) or does the database also has to be multilingual?
b) Are you planning on having a language  "conversion" database or just program static text pages for each of the different languages?

First remember, that the field names on tables do not have to be multi-lingual as they are used internally.

If I was to do multi-lingual, I would design the database in such a way that there are foreign keys to any and all "text" data.  You can then have multiple tables for each language .  For example: user_EN, user_ES, user_FR.  Or you could have a table field called "language" to select which records to retrieve and have a session variable with the Language code.

Do something similar for your /view/ forms.  Use [partials](https://guides.cfwheels.org/cfwheels-guides/displaying-views-to-users/partials) with names such as /view/user/index_EN.cfm, /view/user/index_ES.cfm, /view/user/index_FR.cfm,  Then in /view/user/index.cfm, use the language session variable to append to the [renderPartial() ](https://api.cfwheels.org/v2.4/controller.renderpartial.html) "name" argument.  

The hardest part would be cfwheels errors.  You would have to modify the /events/onerror.cfm and /wheels/events/onerror/*.cfm to handle the output error page of different languages.  As part of /events/onrequeststart, I would preload an application variable with ALL 200-or-so cfwheels error messages from a database. and their language conversions.  I would then change ALL  the  "throw()" error messages in /cfwheels/ to be json data, that /events/onerror.cfm can translate using a pre-loaded database or file. For example; in /wheels/miscellaneous.cfm function sendFile() you could change 
```
Throw(
type = "Wheels.FileNotFound",
message = "A file could not be found.",
extendedInfo = "Make sure a file with the name `#local.file#` exists in the `#local.directory#` folder."
);
```
to

```
Throw(
type = "Wheels.FileNotFound",
message = "{msg_num: 1234, msg_en:'A file could not be found.'} ",
extendedInfo = "{msg_num: 1235, msg_en:"`~~d1~~` exists in the '~~d2~~' folder."
, d1:'#local.file#', d2:'`#local.directory#`}
);

```
Hope that helps.  Dan G.

**Reply by edthetechie** [reply]
Thanks Dan, I hadn't even thought about error messages. The database will contain questions and I'm thinking of a field for each language or a reference to a static language file that gets loaded. Database would be better because the client needs to be able to change the text relatively easily. We haven't taken on the project yet, I've just been asked if we can build it so I'm doing an information gathering exercise. Your reply was most helpful in that regard

**Comment by bpamiri** [comment]
Hi guys, not sure if this is the best way to handle this but this is what we do on one of our projects. I have language files that have a token matched to the text to be used. Then I have these language files translated by native speaking people for each language we want to support. The token in each translation file stays the same but the text would be translated into the target language. 

We load the particular language file either based on geo tagging of the IP or from the user's preferences on a language drop down menu. 

The actual language files are structured as a CFML code that loads a structure into the application scope. So the tokens and their corresponding text are now available in the applications scope. Then we have a few functions that we use in our view templates, where ever we would normally have static text, we make a call out to the function and pass the token we want. The response from the function is the string in whatever language that was loading into the application scope. 

Here is a sampling of a trimmed down version of the English file.

<cfprocessingdirective pageEncoding="utf-8">
<cfscript>

  application.translations.us = {
    chooseLanguageText: "Choose your preferred language below"
    ,english: "English"
    ,language: "Language"
    ,russian: "Russian"
    ,spanish: "Spanish"
    ,submit: "Submit"
 };

</cfscript>

Here is the same sample in the Spanish file:
<cfprocessingdirective pageEncoding="utf-8">
<cfscript>

  application.translations.es = {
    chooseLanguageText: "Seleccione abajo su lenguaje preferido"
    ,english: "Inglés"
    ,language: "Lenguaje"
    ,russian: "Ruso"
    ,spanish: "Español"
    ,submit: "Someter"
 };

</cfscript>


You see the tokens are the same in both files but the string is different. The function that looks up the token is bellow:
  function translate(required string key) {
    if (StructKeyExists(application.translations[session.language], arguments.key)) {
      local.rv = application.translations[session.language][arguments.key];
    } else if (StructKeyExists(application.translations.us, arguments.key)) {
      local.rv = application.translations.us[arguments.key];
    } else {
      local.rv = "";
    }
    return local.rv;
  }

And the way it would be used in a view template would be:
#translate("language")#

This keeps all these calls out of the database, keeps your translations texts in individual language files, and allows you to add additional languages in the future.

**Reply by edthetechie** [reply]
Thanks Peter, that looks like a really great way to build the translation modules for the static text on the site which don't change much. This site will have a massive multi-stage form on it which the client wants to be able to control themselves so that means database driving the questions. I'm still trying to figure out the best way to translate those, so far my ideas are:
1. Use multiple fields in the "questions" and "answers" database tables for each language.
2. Have separate questions/answers for each language (probably not the best way for the answers)
3. Reference a code in the database and link that to a languages file somehow

**Comment by dbelanger** [comment, marked as answer]
Hey Ed,

I'm French Canadian and have been doing multi-language sites for a while now (mostly English/French) and what Peter and Dan have both said are part of a global strategy for i18n (the short form of internationalization).  I spoke about this at ITB 2019 but sadly that talk is behind a paywall...   If you search for i18n, you'll be hit with a wealth of information.

There are really 3 areas that need to be addressed when thinking multi-language:

1. The UI (buttons, field labels, page titles, menu items....)
2. Content (anything that would require a text editor like your questions and answers)
3. Data storage (really just a focus on charset/collation)

If you're going to be content focused, don't hesitate to use a CMS.  I've used MuraCMS and now MasaCMS.  I've also used WordPress (even though it's PHP based, it's hard to beat from a dev and user perspective) as they're already i18n ready.

Summarizing an hour's talk into a few points and following the strategy I mentioned above, this is pretty much what we do in a nutshell:

1. UI - Use a combined code to identify the language and the country (like EN-US).  Stick to the ISO standards ([ISO-3166 and ISO-639](https://docs.oracle.com/cd/E13214_01/wli/docs92/xref/xqisocodes.html)) as it helps a LOT with Google results if you include it in your URL.  Microsoft actually does this very, very well.  Example: [https://www.microsoft.com/fr-ca/](https://www.microsoft.com/fr-ca/)
2. UI - Just like Peter said, the combined language-country code should also be used as a key in a struct for your UI.  There aren't any wheels plugins to help with but honestly, you don't really need one either.  The only difference between Peter and I is that I use JSON files for our language translations.  These JSON files are super easy to load into memory onApplicationStart as application-level structs and are used (and re-used!) site-wide.  Bonus - also super easy to edit manually now with CommandBox if you need to on the fly!!!  We use a structure like this:
`{
  "buttons": {
    {
      "save": {
        "EN-CA" : "Save",
        "EN-FR" : "Sauvegarder",
        "ES-AR" : "Guardar"
      }
    }
  }
}`
3. UI - You'll need an admin page to add language.  Once you add a language, it's very handy for your translator to have a single page with side-by-side columns (language 1 and language 2) for editing.  
4. DB - Make sure you chose the correct charset and collation for your DB from the start!  If you're sticking to the Latin languages, for the most part you can't go wrong with utf8mb4 / utf8mb4_general_ci (in MySQL anyway) which is an extended version of UTF-8 and really has you covered!
5. DB - That same language-country code mentioned in point 2 should be used in a "languageCode" column in your DB.  That's why keeping it to an ISO standard makes everything so easy to find!  
6. Code - Just like the DB's charset, page encoding matters a great deal too!!!  Make sure your files are encoded as UTF-8 (VSCode does this by default) and if you use CF files for your views, always add `<cfprocessingdirective pageencoding="utf-8">` at the top for added security.

To comment on the first point you made when answering Peter, I would always recommend a separate database row per language (not an extra column as you mentioned above) as it gives you far more choices.  Your structure always stays the same and it's easier to map things directly to objects.  I'm also a big fan of composite keys to ensure only one entry per language. In your case, it may be something like question-language (1,EN-CA).  So a database search can return all language options for a single question OR a single question in the specified language.  You can even add your own custom findQuestionByLanguage, customizing the wheels ORM find with those parameters.  

I hope that helps!

**Reply by edthetechie** [reply]
Thank you for that really detailed reply, Mura is a good suggestion but I'm not sure how well it will handle the complex form I need to build. Having said that I will take a look at it before I write off that option. As to the rest of your reply, it's all extremely helpful and gives me a lot to think about. It's always quite daunting taking on a project when there are elements that you've never done before. You made it sound quite easy though!

Thanks to all of you for the replies, so far :-)

**Reply by dbelanger** [reply]
For CMS systems in the CF world, MuraCMS was really good for us (it uses the FW1 framework) but we used it primarily for non-profit sites so when they went from OSS to paid, we switched to MasaCMS.  PresideCMS is also really well done and uses the ColdBox framework as does Ortus' very own ContentBox CMS (not bad but very bare bones).  

There was a project some years ago for a wheels-based CMS called [WheelieCMS](https://github.com/timsayshey/wheelie) but I haven't heard any feedback on it.

**Reply by edthetechie** [reply]
To be honest I've never used a CF CMS, I tend to build them bespoke from the ground up. I can see the advantages though. I looked at your suggestions and need to spend time with them. Preside in particular looks quite interesting

**Comment by edthetechie** [comment]
Thanks very much to @bpamiri and @dbelanger for the very helpful replies, I've marked the final reply as the answer because it contained elements of the other replies but both of you pretty much answered the question. Now I need to decide whether to take on this (quite daunting) project :-)
